File: packages/rust/src/types/memory.rs
//! Memory types for elizaOS
//!
//! Contains Memory, MemoryMetadata, and related types for storing agent memories.

use super::primitives::{Content, UUID};
use serde::{Deserialize, Serialize};

/// Memory type enumeration for built-in memory types
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum MemoryType {
    /// Represents a whole document or a large piece of text
    Document,
    /// A chunk or segment of a document
    Fragment,
    /// A conversational message
    #[default]
    Message,
    /// A descriptive piece of information
    Description,
    /// Custom memory type
    Custom,
}

/// Defines the scope of a memory
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum MemoryScope {
    /// Accessible to multiple entities
    Shared,
    /// Private to a single entity
    Private,
    /// Scoped to a specific room
    #[default]
    Room,
}

/// Base interface for all memory metadata types
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BaseMetadata {
    /// The kind of memory
    #[serde(rename = "type")]
    pub memory_type: String,
    /// Source of the memory
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source: Option<String>,
    /// Source entity ID
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_id: Option<UUID>,
    /// Visibility scope
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scope: Option<MemoryScope>,
    /// Timestamp in milliseconds
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<i64>,
    /// Tags for categorization
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
}

/// Document-specific metadata
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentMetadata {
    /// Base metadata fields
    #[serde(flatten)]
    pub base: BaseMetadata,
}

/// Fragment-specific metadata
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FragmentMetadata {
    /// Base metadata fields
    #[serde(flatten)]
    pub base: BaseMetadata,
    /// Parent document ID
    pub document_id: UUID,
    /// Position in document
    pub position: i32,
}

/// Message-specific metadata
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageMetadata {
    /// Base metadata fields
    #[serde(flatten)]
    pub base: BaseMetadata,
}

/// Description-specific metadata
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DescriptionMetadata {
    /// Base metadata fields
    #[serde(flatten)]
    pub base: BaseMetadata,
}

/// Union type for all memory metadata
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum MemoryMetadata {
    /// Document metadata
    Document(DocumentMetadata),
    /// Fragment metadata
    Fragment(FragmentMetadata),
    /// Message metadata
    Message(MessageMetadata),
    /// Description metadata
    Description(DescriptionMetadata),
    /// Custom metadata (catch-all)
    Custom(serde_json::Value),
}

impl Default for MemoryMetadata {
    fn default() -> Self {
        MemoryMetadata::Custom(serde_json::Value::Object(serde_json::Map::new()))
    }
}

/// Represents a stored memory/message
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Memory {
    /// Optional unique identifier
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<UUID>,
    /// Associated entity ID (user)
    pub entity_id: UUID,
    /// Associated agent ID
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_id: Option<UUID>,
    /// Creation timestamp in milliseconds since epoch
    #[serde(skip_serializing_if = "Option::is_none")]
    pub created_at: Option<i64>,
    /// Memory content
    pub content: Content,
    /// Embedding vector for semantic search
    #[serde(skip_serializing_if = "Option::is_none")]
    pub embedding: Option<Vec<f32>>,
    /// Associated room ID
    pub room_id: UUID,
    /// Associated world ID
    #[serde(skip_serializing_if = "Option::is_none")]
    pub world_id: Option<UUID>,
    /// Whether memory is unique (used to prevent duplicates)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub unique: Option<bool>,
    /// Embedding similarity score (set when retrieved via search)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub similarity: Option<f64>,
    /// Metadata for the memory
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<MemoryMetadata>,
}

impl Memory {
    /// Create a new memory with the given content
    pub fn new(entity_id: UUID, room_id: UUID, content: Content) -> Self {
        Memory {
            id: Some(UUID::new_v4()),
            entity_id,
            agent_id: None,
            created_at: Some(chrono_timestamp()),
            content,
            embedding: None,
            room_id,
            world_id: None,
            unique: Some(true),
            similarity: None,
            metadata: None,
        }
    }

    /// Create a message memory
    pub fn message(entity_id: UUID, room_id: UUID, text: &str) -> Self {
        let content = Content {
            text: Some(text.to_string()),
            ..Default::default()
        };
        let mut memory = Memory::new(entity_id, room_id, content);
        memory.metadata = Some(MemoryMetadata::Message(MessageMetadata {
            base: BaseMetadata {
                memory_type: "message".to_string(),
                ..Default::default()
            },
        }));
        memory
    }
}

/// Get current timestamp in milliseconds
fn chrono_timestamp() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64
}

/// Specialized memory type for messages with enhanced type checking
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageMemory {
    /// The underlying memory
    #[serde(flatten)]
    pub memory: Memory,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_memory_serialization() {
        let memory = Memory::message(UUID::new_v4(), UUID::new_v4(), "Hello, world!");

        let json = serde_json::to_string(&memory).unwrap();

        // Should use camelCase
        assert!(json.contains("\"entityId\""));
        assert!(json.contains("\"roomId\""));
        assert!(json.contains("\"createdAt\""));
    }

    #[test]
    fn test_memory_deserialization() {
        let json = r#"{
            "entityId": "550e8400-e29b-41d4-a716-446655440000",
            "roomId": "550e8400-e29b-41d4-a716-446655440001",
            "content": {
                "text": "Hello from TypeScript!"
            }
        }"#;

        let memory: Memory = serde_json::from_str(json).unwrap();
        assert_eq!(memory.content.text.unwrap(), "Hello from TypeScript!");
    }
}

File: packages/typescript/src/types/memory.ts
import type { Content, MetadataValue, UUID } from "./primitives";
import type {
  BaseMetadata as ProtoBaseMetadata,
  CustomMetadata as ProtoCustomMetadata,
  DescriptionMetadata as ProtoDescriptionMetadata,
  DocumentMetadata as ProtoDocumentMetadata,
  FragmentMetadata as ProtoFragmentMetadata,
  Memory as ProtoMemory,
  MemoryMetadata as ProtoMemoryMetadataType,
  MessageMetadata as ProtoMessageMetadata,
} from "./proto.js";

/**
 * Memory type enumeration for built-in memory types
 */
export type MemoryTypeAlias = string;

/**
 * Enumerates the built-in types of memories that can be stored and retrieved.
 * - `DOCUMENT`: Represents a whole document or a large piece of text.
 * - `FRAGMENT`: A chunk or segment of a `DOCUMENT`, often created for embedding and search.
 * - `MESSAGE`: A conversational message, typically from a user or the agent.
 * - `DESCRIPTION`: A descriptive piece of information, perhaps about an entity or concept.
 * - `CUSTOM`: For any other type of memory not covered by the built-in types.
 * This enum is used in `MemoryMetadata` to categorize memories and influences how they are processed or queried.
 */
export const MemoryType = {
  DOCUMENT: "document",
  FRAGMENT: "fragment",
  MESSAGE: "message",
  DESCRIPTION: "description",
  CUSTOM: "custom",
} as const;

export type MemoryType = (typeof MemoryType)[keyof typeof MemoryType];
/**
 * Defines the scope of a memory, indicating its visibility and accessibility.
 * - `shared`: The memory is accessible to multiple entities or across different contexts (e.g., a public fact).
 * - `private`: The memory is specific to a single entity or a private context (e.g., a user's personal preference).
 * - `room`: The memory is scoped to a specific room or channel.
 * This is used in `MemoryMetadata` to control how memories are stored and retrieved based on context.
 */
export type MemoryScope = "shared" | "private" | "room";

/**
 * Base interface for all memory metadata types.
 * It includes common properties for all memories, such as:
 * - `type`: The kind of memory (e.g., `MemoryType.MESSAGE`, `MemoryType.DOCUMENT`).
 * - `source`: An optional string indicating the origin of the memory (e.g., 'discord', 'user_input').
 * - `sourceId`: An optional UUID linking to a source entity or object.
 * - `scope`: The visibility scope of the memory (`shared`, `private`, or `room`).
 * - `timestamp`: An optional numerical timestamp (e.g., milliseconds since epoch) of when the memory was created or relevant.
 * - `tags`: Optional array of strings for categorizing or filtering memories.
 * Specific metadata types like `DocumentMetadata` or `MessageMetadata` extend this base.
 */
export interface BaseMetadata
  extends Omit<
    ProtoBaseMetadata,
    "$typeName" | "$unknown" | "type" | "scope" | "timestamp"
  > {
  type: MemoryTypeAlias;
  scope?: MemoryScope;
  timestamp?: number;
}

export interface DocumentMetadata
  extends Omit<ProtoDocumentMetadata, "$typeName" | "$unknown" | "base"> {
  base?: BaseMetadata;
  type?: "document";
}

export interface FragmentMetadata
  extends Omit<ProtoFragmentMetadata, "$typeName" | "$unknown" | "base"> {
  base?: BaseMetadata;
  documentId: UUID;
  position: number;
  type?: "fragment";
}

export interface MessageMetadata
  extends Omit<ProtoMessageMetadata, "$typeName" | "$unknown" | "base"> {
  base?: BaseMetadata;
  type?: "message";
  trajectoryStepId?: string;
  benchmarkContext?: string;
}

export interface DescriptionMetadata
  extends Omit<ProtoDescriptionMetadata, "$typeName" | "$unknown" | "base"> {
  base?: BaseMetadata;
  type?: "description";
}

// MetadataValue is imported from primitives.ts

/**
 * Custom metadata with typed dynamic properties
 */
export interface CustomMetadata
  extends Omit<ProtoCustomMetadata, "$typeName" | "$unknown" | "base"> {
  base?: BaseMetadata;
  type?: "custom";
  /** Custom metadata values - must be JSON-serializable */
  [key: string]: MetadataValue | MemoryTypeAlias | BaseMetadata | undefined;
}

interface MemoryMetadataBase {
  type?: MemoryTypeAlias;
  source?: string;
  scope?: MemoryScope;
  timestamp?: number;
}

export type MemoryMetadata = (
  | DocumentMetadata
  | FragmentMetadata
  | MessageMetadata
  | DescriptionMetadata
  | CustomMetadata
) &
  MemoryMetadataBase;

export type ProtoMemoryMetadata = ProtoMemoryMetadataType;

/**
 * Represents a stored memory/message
 */
export interface Memory
  extends Omit<
    ProtoMemory,
    | "$typeName"
    | "$unknown"
    | "id"
    | "createdAt"
    | "embedding"
    | "metadata"
    | "content"
  > {
  id?: UUID;
  createdAt?: number;
  embedding?: number[];
  metadata?: MemoryMetadata;
  content: Content;
}

/**
 * Specialized memory type for messages with enhanced type checking
 */
export type MessageMemory = Memory;
